File: implementations/java/vanilla/src/main/java/org/cwi/waebric/parser/SiteParser.java
package org.cwi.waebric.parser;

import java.util.List;

import org.cwi.waebric.WaebricKeyword;
import org.cwi.waebric.WaebricSymbol;
import org.cwi.waebric.lexer.token.Token;
import org.cwi.waebric.lexer.token.TokenIterator;
import org.cwi.waebric.lexer.token.WaebricTokenSort;
import org.cwi.waebric.parser.ast.module.site.Mapping;
import org.cwi.waebric.parser.ast.module.site.Mappings;
import org.cwi.waebric.parser.ast.module.site.Path;
import org.cwi.waebric.parser.ast.module.site.Site;

/**
 * module languages/waebric/syntax/Site
 * 
 * @see ModuleParser
 * @author Jeroen van Schagen
 * @date 20-05-2009
 */
class SiteParser extends AbstractParser {

	private final MarkupParser markupParser;
	
	/**
	 * Construct site parser
	 * @param tokens
	 * @param exceptions
	 */
	public SiteParser(TokenIterator tokens, List<SyntaxException> exceptions) {
		super(tokens, exceptions);
		
		// Initialize sub-parser
		markupParser = new MarkupParser(tokens, exceptions);
	}
	

	/**
	 * "site" Mapping "end" -> Site
	 * @throws SyntaxException
	 */
	public Site parseSite() throws SyntaxException {
		current(WaebricKeyword.SITE, "Site begin", "\"site\" Mapping \"end\"");
		Site site = new Site();
		site.setMappings(parseMappings());
		next(WaebricKeyword.END, "Site end", "\"site\" Mapping \"end\"");
		return site;
	}
	
	/**
	 * { Mapping ";" }* -> Mappings
	 * @param mappings
	 */
	public Mappings parseMappings() throws SyntaxException {
		Mappings mappings = new Mappings();
		
		while(tokens.hasNext()) {
			mappings.add(parseMapping());
				
			if(tokens.hasNext()) {
				if(tokens.peek(1).getLexeme().equals(WaebricKeyword.END)) {
					break; // End token reached, quit parsing mappings
				} else {
					// Expect mapping separator ";" between each mapping
					next(WaebricSymbol.SEMICOLON, "Mapping separator \";\"", "Mapping \";\" Mapping");
				}
			}
		}
		
		return mappings;
	}
	
	/**
	 * Path ":" Markup -> Mapping
	 * @param mapping
	 */
	public Mapping parseMapping() throws SyntaxException {
		Mapping mapping = new Mapping();
		mapping.setPath(parsePath());
		next(WaebricSymbol.COLON, "Mapping separator \":\"", "Path \":\" Markup");
		mapping.setMarkup(markupParser.parseMarkup());
		return mapping;
	}
	
	/**
	 * DirName "/" FileName -> Path
	 * FileName -> Path
	 * @throws SyntaxException 
	 */
	public Path parsePath() throws SyntaxException {
		String path = ""; // Determine path type based on look-ahead
		if(tokens.hasNext(2) && tokens.peek(2).getLexeme().equals(WaebricSymbol.SLASH)) {
			path = parseDirectory();
		}

		path += parseFileName();
		return new Path(path);
	}
	
	/**
	 * { PathElement "/" }+ -> Directory
	 * @throws SyntaxException 
	 */
	public String parseDirectory() throws SyntaxException {
		String directory = "";
		
		do {
			if(! directory.equals("")) {
				// Expect slash separator between each path element
				next(WaebricSymbol.SLASH, "Path separator \"/\"", "Path element \"/\" Path element");
				directory += "/";
			}
			
			// Detect period separator for potential file names
			if(tokens.hasNext(2) && tokens.peek(2).getLexeme().equals(WaebricSymbol.PERIOD)) {
				break; // File name detected, break from directory parsing
			}

			Token element = tokens.next(); // Retrieve next path element
			if(isPathElement(element.getLexeme().toString())) {
				directory += element.getLexeme().toString();
			} else {
				// Token does not match path element syntax, thus directory is invalid
				reportUnexpectedToken(element, "Path element", "Identifier without layout");
			}
		} while(tokens.hasNext());
		
		return directory;
	}
	
	/**
	 * Check if a certain lexeme is a path element.
	 * @param lexeme
	 * @return PathElement?
	 */
	public static boolean isPathElement(String lexeme) {
		return ! lexeme.matches("(.* .*)|(.*\t.*)|(.*\n.*)|(.*\r.*)|(.*/.*)|(.*\\..*)|(.*\\\\.*)");
	}
	
	/**
	 * PathElement "." FileExt -> FileName
	 * @param File name
	 */
	public String parseFileName() throws SyntaxException {
		String name = "";
		// TODO: Create path element token
		next(WaebricTokenSort.IDCON, "File name", "Name \".\" Extension");
		name += tokens.current().getLexeme().toString() + "."; // Name
		next(WaebricSymbol.PERIOD, "period", "name \".\" extension");
		next(WaebricTokenSort.IDCON, "File extension", "Name \".\" Extension");
		name += tokens.current().getLexeme().toString(); // Extension
		return name;
	}

}

File: implementations/csharp/vanilla/parser/SiteParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Lexer.Tokenizer;
using Parser.Ast.Site;

namespace Parser
{
    /// <summary>
    /// Parser which parses a site
    /// </summary>
    public class SiteParser : AbstractParser
    {
        #region Private Members

        private MarkupParser markupParser;

        #endregion

        public SiteParser(TokenIterator tokenStream) : base(tokenStream)
        {
            //Create parsers
            markupParser = new MarkupParser(tokenStream);
        }

        /// <summary>
        /// Parse a site
        /// </summary>
        /// <returns></returns>
        public Site ParseSite()
        {
            //Create new site
            Site site = new Site();

            //Parse one or more mappings
            site.AddMappings(ParseMappings());

            //Skip END tag
            NextToken("end", "site end", "end");

            return site;
        }

        /// <summary>
        /// Parse mappings
        /// </summary>
        /// <returns>A list of parsed mappings</returns>
        public List<Mapping> ParseMappings()
        {
            List<Mapping> MappingList = new List<Mapping>();

            while (TokenStream.HasNext()) //Parse mappings
            {
                MappingList.Add(ParseMapping());

                if (TokenStream.HasNext()) //Check if another mapping comes
                {
                    if (MatchValue(TokenStream.Peek(1).GetValue().ToString(), Waebric.WaebricKeyword.END.ToString()))
                    {
                        break; //END found so stop searching
                    }
                    else
                    {
                        NextToken(";", "Mapping ; Mapping", ';'); //Match separator
                    }
                }                
            }

            return MappingList;
        }

        /// <summary>
        /// Parse one mapping
        /// </summary>
        /// <returns>Parsed mapping</returns>
        public Mapping ParseMapping()
        {
            Mapping mapping = new Mapping();
            
            // Parse Path
            mapping.SetPath(ParsePath());
            
            // Skip : symbol
            NextToken(":", "path : mapping", ':');

            // Parse Markup
            mapping.SetMarkup(markupParser.ParseMarkup());
            
            return mapping;
        }

        /// <summary>
        /// Parse a path
        /// </summary>
        /// <returns>Parsed path</returns>
        public Path ParsePath()
        {
            Path path = new Path();
           
            //Determine if we have directories in path or just a filename
            if (TokenStream.Peek(2).GetValue().ToString() == ".")
            {   //Just a single filename to parse
                path.SetFilename(ParseFileName());
            }
            else
            {   //Directory and filename
                path.SetDirectoryName(ParseDirectoryName());
                path.SetFilename(ParseFileName());
            }

            return path;
        }

        /// <summary>
        /// Parse a directoryname
        /// </summary>
        /// <returns>Parsed DirName</returns>
        public DirName ParseDirectoryName()
        {
            DirName directoryName = new DirName();
            
            directoryName.SetDirectory(ParseDirectory());
            
            return directoryName;
        }

        /// <summary>
        /// Parse FileName
        /// </summary>
        /// <returns>Parsed FileName</returns>
        public FileName ParseFileName()
        {
            FileName filename = new FileName();

            //Filename
            NextToken("filename", "filename.ext");
            filename.SetName(new PathElement(CurrentToken.GetValue().ToString()));
            
            //Period (between filename and extension)
            NextToken(".", "filename.ext", '.');
            
            //Extension
            NextToken("extension", "filename.ext");
            filename.SetFileExtension(new FileExt(CurrentToken.GetValue().ToString()));

            return filename;
        }

        /// <summary>
        /// Parse Directory
        /// </summary>
        /// <returns>Parsed Directory</returns>
        public Directory ParseDirectory()
        {
            Directory directory = new Directory();
            
            //Parse path elements
            while (TokenStream.HasNext())
            {
                if(TokenStream.Peek(2).GetValue().ToString() == ".")
                { //End of directory, filename starts here
                    break;
                }
                PathElement element = new PathElement();
                NextToken("directory", "directory/filename.ext", TokenType.IDENTIFIER);
                element.SetPathElement(CurrentToken.GetValue().ToString());
                directory.AddDirectoryElement(element);

                //Skip / or \
                NextToken("/ or \\", "directory/FileName.ext", TokenType.SYMBOL);
            }
            return directory;
        }
    }
}
